### DISCUSSION TITLE
Is it OK to rely on/import `addBasePath` and `delBasePath`?

### DISCUSSION THREAD

**Discussion by WickyNilliams** [vercel/next.js#16047] [Help] [answered]
I have a component that takes a url as a prop, and raises an event when the link inside that component is clicked:

```jsx
import { addBasePath, delBasePath } from "next/dist/next-server/lib/router/router"

// this is all simplified for example purposes
<MyComponent
  url="/some-page"
  onButtonClick={e => router.push(e.detail.url)}
/>
```

Because I don't have access to the internals of this component, I have to manually call `router.push` in `handleButtonClick`. This works fine without a base path.

However, with a base path set in next.config.js, I also need to ensure that the base path is _added_ when passing the url to the component (so that the url rendered in the browser is correct), and that the base path is _removed_ when calling `router.push` (since the router expects bare urls with no base path). So my code ends up looking like:

```jsx
import { addBasePath, delBasePath } from "next/dist/next-server/lib/router/router"

// again this is a contrived example...
// i realise i don't actually need delBasePath here,
// but in other instances i do, so it's included for illustration
<MyComponent
  url={addBasePath("/some-page")}
  onButtonClick={e => router.push(delBasePath(e.detail.url))}
/>
```

So my question is, is it safe to rely on these functions? They're imported from `next/dist/next-server/lib/router/router`.

Or are these considered private internals that are likely to change? If they are not safe to rely on, I wonder if these should be exposed as part of some public API (the router perhaps)?

**Comment by Timer** [comment, marked as answer]
No—they are not safe to use, anything you import from `dist/...` is considered private and may break or change at any time.

You can read from the router object instead:

```tsx
import { useRouter } from 'next/router';

const { basePath } = useRouter();
```

**Reply by WickyNilliams** [reply]
Yes, I was aware of this, but that means the logic for handling base paths is restricted to use in react components because of `useRouter`.  Is there any way I can reliably get the base path (e.g. is there an API for reading from next config?) so I am not dependent on the hook?

What are your thoughts on exposing adding/removing base path as a public API? Anyone using base paths needs to handle this stuff, so it's an essential aspect of that feature. It seems a shame that the logic must be reimplemented or that people (like me!) will accidentally depend on private internals. 

Edit: I unmarked answer just in case answered questions don't get seen

**Reply by WickyNilliams** [reply]
There is an env var `__NEXT_ROUTER_BASEPATH` - but i assume this is also private and subject to change?
